src/radical/pilot/utils/component.py

```diff
@@ -4,301 +4,382 @@
+    will not be pushed to outgoing channels automatically -- to do so, the work
+    method has to call (see call documentation for other options):
 
-                comp = ctype.create(tmp_cfg, session)
-                comp.start()
+        self.advance(thing)
 
-                log.info('%-30s starts %s',  tmp_cfg['owner'], comp.uid)
-                components.append(comp)
+    Until that method is called, the component is considered the sole owner of
+    the 'thing's.  After that method is called, the 'thing's are considered
+    disowned by the component.  If, however, components return from the work
+    methods without calling advance on the given 'thing's, then the component
+    keeps ownership of the 'thing's to advance it asynchronously at a later
+    point in time.  That implies that a component can collect ownership over an
+    arbitrary number of 'thing's over time, and they can be advanced at the
+    component's discretion.
 
-        # components are started -- we return the handles to the callee for
-        # lifetime management
-        return components
+    The component process is a stand-alone daemon process which runs outsude of
+    Python's multiprocessing domain.  As such, it can freely use Python's
+    multithreading (and it extensively does so by default) - but developers
+    should be aware that spawning additional *processes* in this component is
+    discouraged, as Python's process management is not playing well with it's
+    multithreading implementation.
+    '''
 
 
     # --------------------------------------------------------------------------
     #
     def __init__(self, cfg, session):
-        """
+        '''
         This constructor MUST be called by inheriting classes, as it specifies
         the operation mode of the component: components can spawn a child
         process, or not.
 
         If a child will be spawned later, then the child process state can be
-        initialized by overloading the`initialize_child()` method.
-        Initialization for component the parent process is similarly done via
-        `initializale_parent()`, which will be called no matter if the component
-        spawns a child or not.
+        initialized by overloading the`initialize()` method.
 
         Note that this policy should be strictly followed by all derived
         classes, as we will otherwise carry state over the process fork.  That
         can become nasty if the state included any form of locking (like, for
         profiling or locking).
 
-        The symmetric teardown methods are called `finalize_child()` and
-        `finalize_parent()`, for child and parent process, repsectively.
+        The symmetric teardown methods are called `finalize()`.
 
         Constructors of inheriting components *may* call start() in their
         constructor.
-        """
+        '''
 
         # NOTE: a fork will not duplicate any threads of the parent process --
```

Comment:
I would add this note to the main docstring

Reply:
Not sure if that is something I would want an end user to be exposed to - it is an implementation level information, isn't it?  We may have a different audience in mind when thinking about docstrings?  I consider what ends up on `readthedocs.net` to be mostly consumed by users, but don't expect it to be sufficient for RCT  developers by any means.  Do we differ here?

Reply:
Following from my previous comment, I think we agree on the distinction between API and impl. details. I think expanding our view of 'user' including developers would be useful. We have a trove of developer-oriented documentation in the code and I think it would be useful to structure it as part of our documentation.